libraries/AP_Scripting/AP_Scripting_helpers.cpp

```diff
@@ -195,4 +195,275 @@ bool Parameter::set_default(float value)
+        return false;
+    }
+
+    if (payloads.space() == 0) {
+        // already have the max number of messages pending, discard
+        // this one, but return true to say it was for us
+        return true;
+    }
+    Payload payload;
+    payload.data = (uint8_t *)malloc(transfer.payload_len);
+    if (payload.data == nullptr) {
+        // discard messages if we can't allocate
+        // but return true to say it was for us
+        return true;
+    }
+
+    payload.length = transfer.payload_len;
+    payload.node_id = transfer.source_node_id;
+    payload.timestamp = transfer.timestamp_usec;
+#if CANARD_ENABLE_CANFD
+    payload.canfd = transfer.canfd;
+#endif
+
+    uint32_t bits_remaining = transfer.payload_len * 8;
+    uint32_t ofs = 0;
+
+    // decode in bytes, we don't have a canardDecodeBuffer function
+    while (bits_remaining > 0) {
+        uint8_t nbits = MIN(8U, bits_remaining);
+        canardDecodeScalar(&transfer, ofs*8, nbits, false, (void*)&payload.data[ofs]);
+        ofs++;
+        bits_remaining -= nbits;
+    }
+
+    // push to queue
+    payloads.push(payload);
+
+    return true;
+}
+
+/*
+  check for an incoming message
+ */
+int DroneCAN_Handle::check_message(lua_State *L)
+{
+    /*
+      get the receive mutex for this driver
+     */
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+
+    auto *s = h->subscriber;
+
+    if (s == nullptr) {
+        return 0;
+    }
+    Subscriber::Payload payload;
+    /*
+      peek first to avoid a memory leak if lua_pushlstring() faults
+     */
+    if (!s->payloads.peek(payload)) {
+        return 0;
+    }
+
+    lua_pushlstring(L, (char *)payload.data, payload.length);
+
+    s->payloads.pop();
+    free(payload.data);
+
+    lua_pushinteger(L, payload.node_id);
+    *new_uint64_t(L) = payload.timestamp;
+#if CANARD_ENABLE_CANFD
+    lua_pushboolean(L, payload.canfd);
+#else
+    lua_pushboolean(L, false);
+#endif
+
+    if (s->trans_type == CanardTransferTypeResponse) {
+        // request reply removes the subscriber
+        WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
+        delete h->subscriber;
+        h->subscriber = nullptr;
+    }
+    
+    return 4;
+}
+
+/*
+  garbage collect a handle
+ */
+int DroneCAN_Handle::__gc(lua_State *L)
+{
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+    WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
+
+    if (h->subscriber != nullptr) {
+        delete h->subscriber;
+        h->subscriber = nullptr;
+    }
+    return 0;
+}
+
+// lua constructor for DroneCAN_Handle
+int DroneCAN_Handle::new_handle(lua_State *L)
+{
+    lua_Number driver_index = luaL_checknumber(L, 1);
+    const uint64_t sig = *check_uint64_t(L, 2);
+    const uint16_t dtype = luaL_checkinteger(L, 3);
+#if CANARD_ENABLE_CANFD
+    const int args = lua_gettop(L);
+    bool send_canfd = false;
+    if (args > 3) {
+        send_canfd = luaL_checkinteger(L, 4) > 0;
```

Comment:
```suggestion
        send_canfd = (bool)lua_toboolean(L, 4);
```

`luaL_checkinteger` will reject booleans; they do not implicitly convert from numbers like C or Python. `lua_toboolean` is what the binding generator uses for boolean argument, but it treats 0 as `true`.

Reply:
good catch, fixed and tested